### ISSUE TITLE
`jnp.searchsorted` returns different values to `np.searchsorted`

### ISSUE THREAD

**Issue by patrick-kidger** [google/jax#9107]
```python
import numpy as np
import jax.numpy as jnp

x = np.arange(10, dtype=float)
x[4:] = jnp.nan
np.searchsorted(x, 0)  # 0
jnp.searchsorted(x, 0)  # DeviceArray(10, dtype=int32)
```
It doesn't look like `jnp.searchsorted` handles NaNs correctly.

**Comment by jakevdp** [comment]
I think this is expected: for `searchsorted` to work correctly, the array must be sorted in increasing order. But for your input I get this:
```python
>>> x[:-1] < x[1:]
array([ True,  True,  True, False, False, False, False, False, False])
```
which implies the array is not in a sorted order, which means the output is ill-defined. We have not made any effort to ensure that JAX's undefined outputs match numpy's undefined outputs, as it would be quite difficult to do so (side note: even for `sort()`, JAX treats NaNs differently than NumPy; see #6443)

What do you think? Is this something you can work around?

**Comment by jakevdp** [comment]
Thinking about it... we could consider this defined behavior in that `sort()` is defined for `NaN` (and `inf`) entries. We could require that the index returned is that that would leave the array unchanged under JAX's definition of `sort()`, probably by first calling `_float_to_int_for_sort` on the entries, as is done when `sort()` is called on a float array: https://github.com/google/jax/blob/f96761cdc861ed6ecaa3bd1b84ca88b257d1b601/jax/_src/lax/lax.py#L3649

That still would not exactly match NumPy's behavior because of the handling of negative NaNs (cf. #6443) but it would definitely be more sensible than the current behavior. What do you think?

**Comment by jakevdp** [comment]
(side-note: this is quite timely becaue I've been working on `searchsorted` all this afternoon: #9108)

**Comment by patrick-kidger** [comment]
I think your proposal to handle NaNs as per `jnp.sort` sounds like a good one.

This is the choice NumPy has made as well: ["As of NumPy 1.4.0 searchsorted works with real/complex arrays containing nan values. The enhanced sort order is documented in sort."](https://numpy.org/doc/stable/reference/generated/numpy.searchsorted.html) (And it would certainly handle my use case.)

The point about negative NaNs is certainly an oddity - since the concept of a negative NaN isn't technically defined - but agreed that this is still more sensible than the current behaviour.

**Comment by jakevdp** [comment]
Complex version is fixed as part of #9108
